pkg/apis/options/options.go

```diff
@@ -68,28 +69,54 @@ type Options struct {
 
 	// internal values that are set after config validation
 	redirectURL        *url.URL
-	provider           providers.Provider
+	provider           []providers.Provider
```

Comment:
I think what we will want to do is remove this from here and instead, copy the pattern we have been moving to of have a constructor within the package.

Eg. Something call `NewProvidersMap` which takes in the configuration of the providers and returns `map[providerID]providers.Provider`

This is something we've been trying to do to reduce the logic required around the options. Options should be considered as an API so we want to avoid contaminating it with lots of logic like this

In general, what I'm aiming for is a lot of smaller packages with well defined configuration structs that return interfaces, once the proxy has been initialised, the `options` struct should be dropped.

An example of this pattern can be seen with the [pagewriter](https://github.com/oauth2-proxy/oauth2-proxy/blob/master/pkg/app/pagewriter/pagewriter.go), see how all of the config going into it is very specific to the part of the code and we have interfaces and a constructor, that is then called when OAuthProxy is being constructed in the main package

Reply:
In terms of this comment, I think we want #1555 as a prerequisite for this

Reply:
That sounds reasonable - I will keep an eye on this space.